File: bindings/java/umicp-core/src/main/java/com/hivellm/umicp/discovery/ServiceDiscovery.java
package com.hivellm.umicp.discovery;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Service Discovery Manager
 * Manages service registration, discovery, and automatic cleanup
 * of stale services in a peer-to-peer network.
 */
public class ServiceDiscovery {
    private final Map<String, ServiceInfo> services;
    private ServiceInfo localService;
    private int timeoutSeconds;

    /**
     * Construct a new Service Discovery with default timeout
     */
    public ServiceDiscovery() {
        this(60); // Default 60 seconds timeout
    }

    /**
     * Construct with custom timeout
     *
     * @param timeoutSeconds Timeout for stale service detection
     */
    public ServiceDiscovery(int timeoutSeconds) {
        this.services = new ConcurrentHashMap<>();
        this.timeoutSeconds = timeoutSeconds;
        this.localService = null;
    }

    /**
     * Register local service
     *
     * @param service Service information
     */
    public synchronized void registerLocal(ServiceInfo service) {
        if (service == null) {
            throw new IllegalArgumentException("service cannot be null");
        }
        this.localService = service;
        services.put(service.getServiceId(), service);
    }

    /**
     * Register a discovered service
     *
     * @param service Service information
     */
    public void registerService(ServiceInfo service) {
        if (service == null) {
            throw new IllegalArgumentException("service cannot be null");
        }

        ServiceInfo existing = services.get(service.getServiceId());
        if (existing != null) {
            existing.updateLastSeen();
        } else {
            services.put(service.getServiceId(), service);
        }
    }

    /**
     * Unregister a service by ID
     *
     * @param serviceId Service ID to unregister
     * @return true if service was found and removed
     */
    public boolean unregisterService(String serviceId) {
        if (serviceId == null || serviceId.isEmpty()) {
            return false;
        }
        return services.remove(serviceId) != null;
    }

    /**
     * Get service by ID
     *
     * @param serviceId Service ID to lookup
     * @return ServiceInfo, or null if not found
     */
    public ServiceInfo getService(String serviceId) {
        if (serviceId == null || serviceId.isEmpty()) {
            return null;
        }
        return services.get(serviceId);
    }

    /**
     * Get all registered services
     *
     * @return List of all services
     */
    public List<ServiceInfo> getAllServices() {
        return new ArrayList<>(services.values());
    }

    /**
     * Find services by capability
     *
     * @param capability Capability to search for
     * @return List of services with the capability
     */
    public List<ServiceInfo> findByCapability(String capability) {
        if (capability == null || capability.isEmpty()) {
            return new ArrayList<>();
        }

        return services.values().stream()
                .filter(s -> s.hasCapability(capability))
                .collect(Collectors.toList());
    }

    /**
     * Find services by metadata key-value pair
     *
     * @param key Metadata key
     * @param value Metadata value
     * @return List of services matching the metadata
     */
    public List<ServiceInfo> findByMetadata(String key, String value) {
        if (key == null || key.isEmpty() || value == null) {
            return new ArrayList<>();
        }

        return services.values().stream()
                .filter(s -> value.equals(s.getMetadataValue(key)))
                .collect(Collectors.toList());
    }

    /**
     * Find services by name pattern (substring match)
     *
     * @param namePattern Name pattern to search for
     * @return List of services matching the pattern
     */
    public List<ServiceInfo> findByName(String namePattern) {
        if (namePattern == null || namePattern.isEmpty()) {
            return new ArrayList<>();
        }

        return services.values().stream()
                .filter(s -> s.getName().contains(namePattern))
                .collect(Collectors.toList());
    }

    /**
     * Clean up stale services
     * Removes services that haven't been seen within the timeout period.
     *
     * @return Number of services removed
     */
    public int cleanupStaleServices() {
        int removedCount = 0;

        Iterator<Map.Entry<String, ServiceInfo>> iterator = services.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<String, ServiceInfo> entry = iterator.next();
            ServiceInfo service = entry.getValue();

            // Don't remove local service
            if (localService != null && service.getServiceId().equals(localService.getServiceId())) {
                continue;
            }

            if (service.isStale(timeoutSeconds)) {
                iterator.remove();
                removedCount++;
            }
        }

        return removedCount;
    }

    /**
     * Get total number of registered services
     *
     * @return Service count
     */
    public int getServiceCount() {
        return services.size();
    }

    /**
     * Get local service info
     *
     * @return Local service, or null if not registered
     */
    public synchronized ServiceInfo getLocalService() {
        return localService;
    }

    /**
     * Set timeout for stale service detection
     *
     * @param timeoutSeconds Timeout in seconds
     */
    public synchronized void setTimeout(int timeoutSeconds) {
        if (timeoutSeconds < 0) {
            throw new IllegalArgumentException("timeoutSeconds must be >= 0");
        }
        this.timeoutSeconds = timeoutSeconds;
    }

    /**
     * Get current timeout setting
     *
     * @return Timeout in seconds
     */
    public int getTimeout() {
        return timeoutSeconds;
    }

    /**
     * Announce service (placeholder for network announcement)
     *
     * @param service Service to announce
     * @return true if announcement was successful
     */
    public boolean announceService(ServiceInfo service) {
        if (service == null) {
            return false;
        }
        // Placeholder - in full implementation, would broadcast to network
        registerService(service);
        return true;
    }
}

File: cpp/src/service_discovery.cpp
/**
 * @file service_discovery.cpp
 * @brief Implementation of Service Discovery for UMICP
 */

#include "service_discovery.h"
#include <algorithm>

namespace umicp {

// ============================================================================
// ServiceInfo Implementation
// ============================================================================

ServiceInfo::ServiceInfo(const std::string& service_id,
                         const std::string& name,
                         const std::string& address)
    : service_id_(service_id)
    , name_(name)
    , address_(address)
    , last_seen_(std::chrono::system_clock::now())
    , version_("1.0.0")
{
}

void ServiceInfo::addCapability(const std::string& capability) {
    // Only add if not already present
    if (std::find(capabilities_.begin(), capabilities_.end(), capability) == capabilities_.end()) {
        capabilities_.push_back(capability);
    }
}

void ServiceInfo::addMetadata(const std::string& key, const std::string& value) {
    metadata_[key] = value;
}

bool ServiceInfo::hasCapability(const std::string& capability) const {
    return std::find(capabilities_.begin(), capabilities_.end(), capability) != capabilities_.end();
}

std::string ServiceInfo::getMetadataValue(const std::string& key) const {
    auto it = metadata_.find(key);
    if (it != metadata_.end()) {
        return it->second;
    }
    return "";
}

void ServiceInfo::updateLastSeen() {
    last_seen_ = std::chrono::system_clock::now();
}

bool ServiceInfo::isStale(int timeout_seconds) const {
    auto now = std::chrono::system_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - last_seen_);
    return elapsed.count() > timeout_seconds;
}

// ============================================================================
// ServiceDiscovery Implementation
// ============================================================================

ServiceDiscovery::ServiceDiscovery()
    : timeout_seconds_(60)
{
}

ServiceDiscovery::ServiceDiscovery(int timeout_seconds)
    : timeout_seconds_(timeout_seconds)
{
}

void ServiceDiscovery::registerLocal(const ServiceInfo& service) {
    std::lock_guard<std::mutex> lock(mutex_);
    local_service_ = std::make_shared<ServiceInfo>(service);
    services_[service.getServiceId()] = local_service_;
}

void ServiceDiscovery::registerService(const ServiceInfo& service) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = services_.find(service.getServiceId());

    if (it != services_.end()) {
        // Update existing service
        it->second = std::make_shared<ServiceInfo>(service);
        it->second->updateLastSeen();
    } else {
        // Add new service
        services_[service.getServiceId()] = std::make_shared<ServiceInfo>(service);
    }
}

bool ServiceDiscovery::unregisterService(const std::string& service_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = services_.find(service_id);

    if (it != services_.end()) {
        services_.erase(it);
        return true;
    }

    return false;
}

std::shared_ptr<ServiceInfo> ServiceDiscovery::getService(const std::string& service_id) const {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = services_.find(service_id);

    if (it != services_.end()) {
        return it->second;
    }

    return nullptr;
}

std::vector<std::shared_ptr<ServiceInfo>> ServiceDiscovery::getAllServices() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::shared_ptr<ServiceInfo>> result;
    result.reserve(services_.size());

    for (const auto& pair : services_) {
        result.push_back(pair.second);
    }

    return result;
}

std::vector<std::shared_ptr<ServiceInfo>> ServiceDiscovery::findByCapability(
    const std::string& capability) const {

    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::shared_ptr<ServiceInfo>> result;

    for (const auto& pair : services_) {
        if (pair.second->hasCapability(capability)) {
            result.push_back(pair.second);
        }
    }

    return result;
}

std::vector<std::shared_ptr<ServiceInfo>> ServiceDiscovery::findByMetadata(
    const std::string& key, const std::string& value) const {

    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::shared_ptr<ServiceInfo>> result;

    for (const auto& pair : services_) {
        if (pair.second->getMetadataValue(key) == value) {
            result.push_back(pair.second);
        }
    }

    return result;
}

std::vector<std::shared_ptr<ServiceInfo>> ServiceDiscovery::findByName(
    const std::string& name_pattern) const {

    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::shared_ptr<ServiceInfo>> result;

    for (const auto& pair : services_) {
        if (pair.second->getName().find(name_pattern) != std::string::npos) {
            result.push_back(pair.second);
        }
    }

    return result;
}

int ServiceDiscovery::cleanupStaleServices() {
    std::lock_guard<std::mutex> lock(mutex_);
    int removed_count = 0;

    auto it = services_.begin();
    while (it != services_.end()) {
        if (it->second->isStale(timeout_seconds_)) {
            // Don't remove local service
            if (local_service_ && it->second->getServiceId() == local_service_->getServiceId()) {
                ++it;
                continue;
            }

            it = services_.erase(it);
            removed_count++;
        } else {
            ++it;
        }
    }

    return removed_count;
}

size_t ServiceDiscovery::getServiceCount() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return services_.size();
}

std::shared_ptr<ServiceInfo> ServiceDiscovery::getLocalService() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return local_service_;
}

void ServiceDiscovery::setTimeout(int timeout_seconds) {
    std::lock_guard<std::mutex> lock(mutex_);
    timeout_seconds_ = timeout_seconds;
}

bool ServiceDiscovery::announceService(const ServiceInfo& service) {
    // Placeholder for network announcement
    // In a full implementation, this would broadcast the service
    // information to the network (e.g., via multicast or central registry)

    // For now, just register it locally
    registerService(service);
    return true;
}

} // namespace umicp

